designs/s3/design-vscode-s3.md

```diff
@@ -0,0 +1,262 @@
+    capabilties. (Note: S3 only supports prefix-filtering.)
+  - The `Filter...` action prompts user for a prefix string.
+    - Pre-populate the prompt with the current active filter pattern, if any.
+  - Store the current active filter per-folder.
+  - When user applies a filter, rebuild the entire subtree of the folder
+    (discard old local state). This also applies to the `More...` action:
+    clicking `More...` fetches the next _filtered_ page of results.
+  - Indicate the current active filter by showing an icon/sigil next to the folder.
+    - If user mouse-hovers the folder, show a tooltip like `Filter: "foo bar"`.
+  - Mouse-hover on a bucket/folder reveals [inline](https://code.visualstudio.com/api/extension-guides/tree-view#view-actions)
+    `Filter...` button (so user does not need to dig into the context-menu).
+  - Context-menu includes `Filter...` menu item
+
+### S3 root node
+
+The S3 root node is the "meta node" which lists available buckets in the
+current region.
+
+- Toolkit toolbar:
+  <br><img src="design-vscode-s3-toolbar.png" width="200"/>
+  - (P0) When any S3 node (bucket/folder/object) is selected, Toolkit toolbar
+    enables an `Upload...` button.
+    - If a non-folder object is selected, `Upload...` targets the object's parent folder.
+- S3 root node context-menu:
+  - (P0) `New bucket`
+    - Invokes the `AWS: New bucket` VSCode command which prompts for Bucket name.
+    - Public access is blocked by default. To enable public access, user must
+      use the AWS web console.
+
+### S3 buckets
+
+Buckets are listed as children of the S3 root node.
+
+- Bucket node context-menu:
+  - (P0) `Copy ARN`
+  - (P0) `Copy name`
+  - (P0) `New folder...`
+  - (P0) `Upload...`
+    - Shows VSCode-standard "Open file..." dialog.
+    - Shows a progress dialog using VSCode conventions.
+    - When upload succeeds or fails, shows a toaster summarizing which uploads succeeded and failed.
+    - (P1) Allows multiple files and folders to be uploaded. Creates S3 folders
+      as needed to match the hierarchy of uploaded folders.
+    - (P2) [Auto-detects Content Type ](https://github.com/aws/aws-toolkit-visual-studio/issues/64)
+  - (P0) `Upload to parent folder...`
+    - Works like `Upload...`, except targeting the parent of the current item.
+      This allows the user to avoid needing to navigate to the parent folder.
+  - _Destructive operations (separator):_
+    - (P1) `Delete bucket`
+      - "Super confirm": similar to AWS web console, show a prompt which
+        requires typing the word "Delete".
+
+### S3 objects
+
+S3 objects and folders are listed as children of their respective parent bucket/folder.
+
+- (P0) Clicking an object in the tree selects it, but does not perform any action.
+- (P1) Double-clicking an item downloads it and opens it in readonly mode (see `Edit` action).
+  - If item is 4MB+, prompt user to ask if download should begin.
```

Comment:
Seems a bit arbitrary. Would this be surfaced to the user somehow?

Reply:
It is arbitrary.